### DISCUSSION TITLE
Sub menu example in README segmentation fault

### DISCUSSION THREAD

**Discussion by lu-maca** [daniele77/cli#241] [Q&A]
Hi,

I'm having some trouble with the example in README. This compiles correctly but gives a segfault when executed.

```cpp

#include "cli/cli.h"
#include "cli/clifilesession.h"

using namespace cli;
using namespace std;

int main() {
  // setup cli

  auto rootMenu = make_unique<Menu>("myprompt");

  auto menuA = make_unique<Menu>("a_prompt");
  rootMenu->Insert(std::move(menuA));

  auto menuAA = make_unique<Menu>("aa_prompt");
  menuA->Insert(std::move(menuAA));

  auto menuAB = make_unique<Menu>("ab_prompt");
  menuA->Insert(std::move(menuAB));

  auto menuAC = make_unique<Menu>("ac_prompt");
  menuA->Insert(std::move(menuAC));

  auto menuACA = make_unique<Menu>("aca_prompt");
  menuAC->Insert(std::move(menuACA));

  auto menuB = make_unique<Menu>("b_prompt");
  rootMenu->Insert(std::move(menuB));

  auto menuBA = make_unique<Menu>("ba_prompt");
  menuB->Insert(std::move(menuBA));

  auto menuBB = make_unique<Menu>("bb_prompt");
  menuB->Insert(std::move(menuBB));

  Cli cli(std::move(rootMenu));
  // global exit action
  cli.ExitAction(
      [](auto &out) { out << "Goodbye and thanks for all the fish.\n"; });

  CliFileSession input(cli);
  input.Start();

  return 0;
}

```

A way of making it work is the following:
```cpp
int main() {
  // setup cli

  auto rootMenu = make_unique<Menu>("myprompt");

  auto menuA = make_unique<Menu>("a_prompt");
  auto menuAA = make_unique<Menu>("aa_prompt");
  auto menuAB = make_unique<Menu>("ab_prompt");
  auto menuAC = make_unique<Menu>("ac_prompt");
  auto menuACA = make_unique<Menu>("aca_prompt");

  auto menuB = make_unique<Menu>("b_prompt");
  auto menuBA = make_unique<Menu>("ba_prompt");
  auto menuBB = make_unique<Menu>("bb_prompt");
  
  menuAC->Insert(std::move(menuACA));
  
  menuB->Insert(std::move(menuBB));
  menuB->Insert(std::move(menuBA));

  menuA->Insert(std::move(menuAA));
  menuA->Insert(std::move(menuAB));
  menuA->Insert(std::move(menuAC));

  rootMenu->Insert(std::move(menuA));
  rootMenu->Insert(std::move(menuB));

  Cli cli(std::move(rootMenu));
  // global exit action
  cli.ExitAction(
      [](auto &out) { out << "Goodbye and thanks for all the fish.\n"; });

  CliFileSession input(cli);
  input.Start();

  return 0;
}

```

I'm not sure why, unfortunately I don't have much time to study the problem. Hope this can help.

Thanks for you work,
Luca

**Comment by daniele77** [comment]
Hi Luca,

Thank you for bringing this to our attention. Upon review, the example you mentioned is not included in the README.md file or any of the library's provided examples.

It appears that the issue stems from the use of a unique_ptr variable that has been previously moved (using std::move). Once a variable has been moved, it enters an undefined state, and attempting to use it will result in a crash.

In your second example, rearranging the statements so that variable moves occur as the last action resolves the issue. This ensures that variables are only moved after they have been fully utilized.

Please let me know if you have any further questions or concerns.

**Reply by lu-maca** [reply]
Ciao Daniele,

actually the example is in the README!
![immagine](https://github.com/daniele77/cli/assets/65252677/4d5713cf-e757-40be-baef-936b1eae5d5d)

Thanks for the explanation, it makes sense :)

Luca

**Reply by daniele77** [reply]
Sorry, you're right.
That section of the README file is completely wrong.
I'm going to fix the file immediately.
Thank you very much for telling me.

**Reply by lu-maca** [reply]
Thank you for the library!
